lib/trino-hive-formats/src/main/java/io/trino/hive/formats/rcfile/RcFileReader.java

```diff
@@ -144,100 +143,107 @@ private RcFileReader(
-        columns = new Column[columnCount];
-        for (Entry<Integer, Type> entry : readColumns.entrySet()) {
-            if (entry.getKey() < columnCount) {
-                ColumnEncoding columnEncoding = encoding.getEncoding(entry.getValue());
-                columns[entry.getKey()] = new Column(columnEncoding, decompressor);
+            // initialize columns
+            verify(columnCount <= MAX_COLUMN_COUNT, "Too many columns (%s) in RCFile %s", columnCountString, inputFile.location());
+            columns = new Column[columnCount];
+            for (Entry<Integer, Type> entry : readColumns.entrySet()) {
+                if (entry.getKey() < columnCount) {
+                    ColumnEncoding columnEncoding = encoding.getEncoding(entry.getValue());
+                    columns[entry.getKey()] = new Column(columnEncoding, decompressor);
+                }
             }
-        }
 
-        // read sync bytes
-        syncFirst = input.readLong();
-        validateWrite(validation -> validation.getSyncFirst() == syncFirst, "Unexpected sync sequence");
-        syncSecond = input.readLong();
-        validateWrite(validation -> validation.getSyncSecond() == syncSecond, "Unexpected sync sequence");
-
-        // seek to first sync point within the specified region, unless the region starts at the beginning
-        // of the file.  In that case, the reader owns all row groups up to the first sync point.
-        if (offset != 0) {
-            // if the specified file region does not contain the start of a sync sequence, this call will close the reader
-            long startOfSyncSequence = ReadWriteUtils.findFirstSyncPosition(inputFile, offset, length, syncFirst, syncSecond);
-            if (startOfSyncSequence < 0) {
-                closeQuietly();
-                return;
+            // read sync bytes
+            syncFirst = input.readLong();
+            validateWrite(validation -> validation.getSyncFirst() == syncFirst, "Unexpected sync sequence");
+            syncSecond = input.readLong();
+            validateWrite(validation -> validation.getSyncSecond() == syncSecond, "Unexpected sync sequence");
+
+            // seek to first sync point within the specified region, unless the region starts at the beginning
+            // of the file.  In that case, the reader owns all row groups up to the first sync point.
+            if (offset != 0) {
+                // if the specified file region does not contain the start of a sync sequence, this call will close the reader
+                long startOfSyncSequence = ReadWriteUtils.findFirstSyncPosition(inputFile, offset, length, syncFirst, syncSecond);
+                if (startOfSyncSequence < 0) {
+                    closeQuietly();
+                    return;
+                }
+                input.seek(startOfSyncSequence);
             }
-            input.seek(startOfSyncSequence);
+        }
+        catch (Exception e) {
+            closeQuietly();
```

Comment:
we're calling close from within constructor, in pretty undetermined state.
the close main fail with IOException (being caught), but it can also fail with any other exception (eg NPE if some field not assigned yet).

wrap with try-catch. add any exception as suppressed to e (unless it's `e` itself, rethrown again)

```
try {
 close();
catch (Exception closeException) {
 if (e!=closeException) {
  e.addSuppressed(closeException);
 }
}
```